server/app/controllers/applicant/UpsellController.java

```diff
@@ -200,13 +201,14 @@ public CompletionStage<Result> download(
         return CompletableFuture.completedFuture(unauthorized("Invalid credentials"));
       }
     }
-    Optional<Application> applicationMaybe =
-        applicationService.getApplication(applicationId, program);
-    if (applicationMaybe.isEmpty()) {
+    CompletableFuture<Optional<Application>> applicationMaybe =
+        applicationService.getApplication(applicationId, program.join()).toCompletableFuture();
```

Comment:
**issue (blocking):** adding `join()` here is making this previously async code blocking

Reply:
moved the join to the service class. We do need to check the application if it has the right program name.